File: backends/ts-hono/src/repositories/notificationRepo.ts
import { and, count, eq, gt, lt, or } from "drizzle-orm";
import { db } from "../db/client.js";
import { notifications, type Notification } from "../db/schema.js";

export interface NotificationPage {
  items: Notification[];
  nextCursor: string | null;
}

function encodeCursor(ts: number, id: number): string {
  return Buffer.from(`${ts},${id}`).toString("base64");
}

function decodeCursor(cursor: string): { ts: number; id: number } {
  const [ts, id] = Buffer.from(cursor, "base64").toString().split(",");
  return { ts: parseInt(ts!), id: parseInt(id!) };
}

export async function listPage(staffId: number, cursor: string | undefined, limit: number): Promise<NotificationPage> {
  let cond = eq(notifications.staffId, staffId);

  if (cursor) {
    const { ts, id } = decodeCursor(cursor);
    const curDate = new Date(ts * 1000);
    cond = and(
      cond,
      or(
        lt(notifications.createdAt, curDate),
        and(eq(notifications.createdAt, curDate), lt(notifications.id, id)),
      )!,
    )!;
  }

  const rows = await db.select().from(notifications)
    .where(cond)
    .orderBy(notifications.createdAt, notifications.id)
    .limit(limit + 1);

  let nextCursor: string | null = null;
  if (rows.length > limit) {
    rows.splice(limit);
    const last = rows[rows.length - 1]!;
    const ts = Math.floor((last.createdAt?.getTime() ?? 0) / 1000);
    nextCursor = encodeCursor(ts, last.id);
  }

  return { items: rows, nextCursor };
}

export async function countNotifications(staffId: number): Promise<{ unread: number; total: number }> {
  const [totalRow] = await db.select({ cnt: count() }).from(notifications).where(eq(notifications.staffId, staffId));
  const [unreadRow] = await db.select({ cnt: count() }).from(notifications).where(and(eq(notifications.staffId, staffId), eq(notifications.read, false)));
  return { total: totalRow?.cnt ?? 0, unread: unreadRow?.cnt ?? 0 };
}

export async function bulkMarkRead(executorId: number, ids: number[], allFlag: boolean): Promise<number> {
  const cond = allFlag
    ? and(eq(notifications.staffId, executorId), eq(notifications.read, false))!
    : and(eq(notifications.staffId, executorId), eq(notifications.read, false))!;

  const targets = await db.select().from(notifications).where(cond);
  const filtered = allFlag ? targets : targets.filter(n => ids.includes(n.id));
  const updated = filtered.length;
  if (updated > 0) {
    for (const n of filtered) {
      await db.update(notifications).set({ read: true, updatedAt: new Date() }).where(eq(notifications.id, n.id));
    }
  }
  return updated;
}

export async function insertNotification(data: typeof notifications.$inferInsert): Promise<void> {
  await db.insert(notifications).values(data);
}

export async function findNotificationById(id: number): Promise<Notification | undefined> {
  const rows = await db.select().from(notifications).where(eq(notifications.id, id)).limit(1);
  return rows[0];
}

export async function patchNotification(id: number, data: Partial<Pick<Notification, "read" | "title" | "body">>): Promise<void> {
  await db.update(notifications).set({ ...data, updatedAt: new Date() }).where(eq(notifications.id, id));
}

File: backends/python-fastapi/app/repositories/notification_repo.py
import base64
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional
from sqlalchemy.orm import Session
from app.models.models import Notification


@dataclass
class NotificationPage:
    items: list[Notification]
    next_cursor: Optional[str]


def _encode_cursor(ts: int, nid: int) -> str:
    raw = f"{ts},{nid}"
    return base64.b64encode(raw.encode()).decode()


def _decode_cursor(cursor: str) -> tuple[int, int]:
    raw = base64.b64decode(cursor.encode()).decode()
    ts_str, id_str = raw.split(",", 1)
    return int(ts_str), int(id_str)


class NotificationRepository:
    def __init__(self, db: Session):
        self.db = db

    def list_page(self, staff_id: int, cursor: Optional[str], limit: int) -> NotificationPage:
        q = self.db.query(Notification).filter(Notification.staff_id == staff_id)
        if cursor:
            ts, nid = _decode_cursor(cursor)
            cur_dt = datetime.fromtimestamp(ts, tz=timezone.utc)
            q = q.filter(
                (Notification.created_at < cur_dt) |
                ((Notification.created_at == cur_dt) & (Notification.id < nid))
            )
        q = q.order_by(Notification.created_at.desc(), Notification.id.desc())
        items = q.limit(limit + 1).all()

        next_cursor = None
        if len(items) > limit:
            items = items[:limit]
            last = items[-1]
            ts = int(last.created_at.replace(tzinfo=timezone.utc).timestamp())
            next_cursor = _encode_cursor(ts, last.id)

        return NotificationPage(items=items, next_cursor=next_cursor)

    def counts(self, staff_id: int) -> tuple[int, int]:
        total = self.db.query(Notification).filter(Notification.staff_id == staff_id).count()
        unread = self.db.query(Notification).filter(
            Notification.staff_id == staff_id, Notification.read == False  # noqa: E712
        ).count()
        return unread, total

    def bulk_mark_read(self, executor_id: int, ids: list[int], all_flag: bool) -> int:
        q = self.db.query(Notification).filter(Notification.staff_id == executor_id)
        if not all_flag:
            q = q.filter(Notification.id.in_(ids))
        count = q.filter(Notification.read == False).count()  # noqa: E712
        q.filter(Notification.read == False).update({"read": True}, synchronize_session=False)  # noqa: E712
        self.db.commit()
        return count

    def store(self, notification: Notification) -> Notification:
        self.db.add(notification)
        self.db.commit()
        self.db.refresh(notification)
        return notification

    def patch(self, notification: Notification, data: dict) -> None:
        allowed = {"read", "title", "body"}
        for key, val in data.items():
            if key in allowed:
                setattr(notification, key, val)
        self.db.commit()

    def find_by_id(self, nid: int) -> Optional[Notification]:
        return self.db.query(Notification).filter(Notification.id == nid).first()
